add support for expanding ${file}. if ${file} is used, reload on completion

jonasmr opened the pull request:
- Add support for expanding ${file}. if ${file} is used, reload the file on completion.
With this you can run a program on the current file, IE if for checking out a file in perforce:

		{
			"name": "p4 edit",
			"type": "shell",
			"command": "p4",
			"args": "edit ${file}",
			"windows": {
				"command": "p4",
				"args": "edit ${file}",
			},
			"show_output_panel": true,
		},
- Fixed an issue where the project file would not be found, if it wasn't located in the first folder of the project.
- The project file parser now removes trailing commas in from the project file, and accepts its, in the same way sublime text itself does.
- The project file parser now prints the line where it fails parsing the project file.

jonasmr committed:
add support for expanding ${file}. if ${file} is used, reload on completion

jonasmr committed:
update non console version of run-task to support @file, and make it print its output to the default console

jonasmr committed:
use project_file_name to find project file. Previous method only worked if the first folder in the project was the folder the project was located in

jonasmr commented:
Hey! Any chance of you considering merging this?

jonasmr committed:
Add support for trailing commas in project file. Add a real error message when json parsing fails

mgutski commented:
> Hey! Any chance of you considering merging this?

Hey Jonas!

Sorry, I completely overlooked your PR. As for the code review, my Open Source backlog is very much full and this particular project is not on my priority list at the moment. Nonetheless, I will try to find some time to go through your changes in the upcoming week or two.

In a meantime, would you mind adding a PR description explaining the rationale behind your changes and giving an example use case for the added feature? While I get the general idea of introducing the `${file}` substitutable variable and I can see it being useful, having more details on what/why/how will certainly help in making sure we are on the same page before I start the review.

P.S. I appreciate the time you took to contribute to this (a little forgotten) project! Perhaps it's time to give it a little TLC.

jonasmr commented:
Hi!

Thanks - I've updated the description now.

-Jonas

mgutski commented on run_task.py:
```diff
@@ -337,17 +368,25 @@ def run(self):
 		self.tasks = []
 
-		project_file = OSUtils.find_file_with_pattern(self.workspace, PROJECT_FILE_NAME_PATTERN)
+		project_file = self.window.project_file_name()
```
This looks to be a breaking change, as the [`project_file_name()`](https://www.sublimetext.com/docs/api_reference.html#sublime.Window.project_file_name) method will only work if there is an opened Sublime project file. Whereas, Run Task was designed to work even if you open the project root directory via `sublime .`. With that said, I think it would be best to use both methods, i.e. call `self.window.project_file_name()` first and fallback to `OSUtils.find_file_with_pattern(self.workspace, PROJECT_FILE_NAME_PATTERN)` in case no project file is opened.

mgutski commented on run_task.py:
```diff
@@ -337,17 +368,25 @@ def run(self):
+
 				try:
-					tasks_json = json.load(fp)
+					json_str = fp.read()
+					# sublime supports trailing commas in the json files, json doesn't. Fix this here
+					
+					pattern = r',\s*([\]}])'
+					json_str_fixed = re.sub(pattern, r'\1', json_str)
+					tasks_json = json.loads(json_str_fixed)
```
There seems to be a much simpler and cleaner solution to this project file parsing issue. We could take advantage of the [`sublime.decode_value(data: str)`](https://www.sublimetext.com/docs/api_reference.html#sublime.decode_value) function and refactor your code to this:
```python
json_str = fp.read()
tasks_json = sublime.decode_value(json_str)
```
As an added bonus, the above code will also gracefully handle the comments in the project file, as per the Sublime API Reference.

mgutski commented on run_task.py:
```diff
@@ -337,17 +368,25 @@ def run(self):
 						return
-				except ValueError:
-					sublime.error_message(ErrorMessage.invalid_json())
+				except ValueError as error:
+					error_message = "JSON decoding error: %s " % str(error)
+					sublime.error_message(error_message)
```
This new error message is now missing a basic information about the package that encountered an error, as well as the name of the malformed JSON file. From the user perspective, how would I know **which file** is it that needs to be fixed? To address that, the code should be changed as follows:
```python
except ValueError as error:
	sublime.error_message(ErrorMessage.invalid_json(str(error)))
```
, and the `invalid_json` method definition should be changed to:
```python
def invalid_json(json_error_message):
	error_message = PLUGIN_NAME + ': Invalid JSON. ' + ErrorMessage.CHECK_CONFIGURATION_FILE
	error_message += '\n\nError message: ' + json_error_message
	return error_message
```
This way we make sure the user gets all the information they need to fix the invalid JSON error. Also, by keeping all the error functions encapsulated in the `ErrorMessage` class, we ensure the consistent error message format across all the errors reported by Run Task.

jonasmr committed:
Updated with suggestions from review

- Fallback to search for project file if no file is loaded
- Use sublime.decode_value to load the project file
- More verbose error message

jonasmr commented:
Thank you for the feedback. I've updated the code now - All your suggestions are added.

mgutski commented on run_task.py:
```diff
@@ -6,7 +6,8 @@
 import shlex
 import threading
-
+import sys
+import re
```
I think these two imports can now be safely removed.

mgutski commented on run_task.py:
```diff
@@ -337,17 +370,24 @@ def run(self):
+		project_file = self.window.project_file_name()
+		if not project_file:
+			project_file = OSUtils.find_file_with_pattern(self.workspace, PROJECT_FILE_NAME_PATTERN)
+
+
```
A single empty line will suffice.

mgutski commented on run_task.py:
```diff
@@ -337,17 +370,24 @@ def run(self):
+
 		if project_file:
 			with open(project_file, "r") as fp:
+
```
Redundant empty line.

mgutski commented on run_task.py:
```diff
@@ -337,17 +370,24 @@ def run(self):
-					tasks_json = json.load(fp)
+					json_str = fp.read()
+					tasks_json = sublime.decode_value(json_str)
+
```
Another redundant empty line.

mgutski commented on run_task.py:
```diff
@@ -205,6 +233,8 @@ def set_show_output_panel(self, value):
 	show_output_panel = property(get_show_output_panel, set_show_output_panel)
 
 	def execute(self, window, cwd):
+		variables = window.extract_variables()
+		file = variables['file']
```
These two variables should ideally be moved closer to the place of use.

mgutski commented on run_task.py:
```diff
@@ -213,9 +243,12 @@ def execute(self, window, cwd):
 				args.extend(self.args)
 			elif type(self.args) is str:
 				args.extend(shlex.split(self.args.strip()))
+			has_file = False
```
Is the introduction of this parameter absolutely necessary? I can see that it is essentially only being used for determining whether to call `self.window.focus_view(self.window.active_view())`, but what is the practical implication of this? I can imagine that if there was some long-running task (which is rather rare case), then it could potentially make sense to retain the focus on an active view once the task finishes. But then again, the [`active_view()`](https://www.sublimetext.com/docs/api_reference.html#sublime.Window.active_view) could easily change in the meantime, at which point the described logic would no longer work as expected. Overall, I would just suggest removing it.

mgutski commented on run_task.py:
```diff
@@ -213,9 +243,12 @@ def execute(self, window, cwd):
 				args.extend(shlex.split(self.args.strip()))
+			has_file = False
 			for idx, arg in enumerate(args):
 				args[idx] = arg.replace(VARIABLE_CWD, cwd)
-			ShellTaskThread(self.name, window, args, cwd, self.show_output_panel).start()
+				has_file = has_file or VARIABLE_FILE in args[idx]
+				args[idx] = arg.replace(VARIABLE_FILE, file)
+			ShellTaskThread(self.name, window, args, cwd, self.show_output_panel, has_file).start()
```
Given my previous comments, if you remove the `has_file` parameter, then this code could be simplified as follows:
```python
	variables = window.extract_variables()
	args = sublime.expand_variables(args, {VARIABLE_CWD: cwd, VARIABLE_FILE: variables[VARIABLE_FILE]})
	ShellTaskThread(self.name, window, args, cwd, self.show_output_panel).start()
```
That also assumes changing the values of `VARIABLE_CWD` and `VARIABLE_FILE` constants to `cwd` and `file` respectively.

mgutski commented on run_task.py:
```diff
@@ -146,14 +151,37 @@ def run(self):
+				except Exception as ex:
+					sublime.error_message(ErrorMessage.task_execution_failed(self.task_name, str(ex)))
+				while True:
+					out_line = process.stdout.readline()
+					if out_line:
+						print(out_line)
+
```
I don't think we should be writing out the output to the Sublime Text console (i.e. the stdout). Depending on the task being run, this could clutter the logs and impede the readability of diagnostic messages. If someone wants to inspect the task's output, then they should set the `show_output_panel` property to `true`.

mgutski commented on run_task.py:
```diff
@@ -146,14 +151,37 @@ def run(self):
+					sublime.error_message(ErrorMessage.task_execution_failed(self.task_name, str(ex)))
+			else:
+				try:
+					si = subprocess.STARTUPINFO()
+					si.dwFlags |= subprocess.STARTF_USESHOWWINDOW
+					si.wShowWindow = subprocess.SW_HIDE # default
```
As per the [docs](https://docs.python.org/3/library/subprocess.html#subprocess.STARTUPINFO), the `STARTUPINFO` class is only available on Windows, so this code will crash on Linux and macOS. At the same time, I believe the `si.wShowWindow = subprocess.SW_HIDE` is needed to prevent the command shell window from opening when running the task on Windows, but please correct me if I'm wrong.

To fix the above, and address my previous comments, I would suggest to rewrite the `run()` method as follows:
```python
if platform.system() == 'Windows':
	si = subprocess.STARTUPINFO()
	si.dwFlags |= subprocess.STARTF_USESHOWWINDOW
	si.wShowWindow = subprocess.SW_HIDE
else:
	si = None

if self.show_output_panel:
	output_panel = OutputPanel(self.window)
	output_panel.show()
	try:
		process = subprocess.Popen(self.args, bufsize=1, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, cwd=self.cwd, universal_newlines=True, startupinfo=si)
	except Exception as ex:
		sublime.error_message(ErrorMessage.task_execution_failed(self.task_name, str(ex)))
		return
	while True:
		out_line = process.stdout.readline()
		if out_line == '' and process.poll() is not None:
			finish_msg = "| Task finished with return code " + str(process.poll()) + " |"
			finish_msg = "\n" + ("-" * len(finish_msg)) + "\n" + finish_msg + "\n" + ("-" * len(finish_msg))
			output_panel.write(finish_msg)
			break
		if out_line:
			output_panel.write(out_line)
else:
	try:
		subprocess.Popen(self.args, cwd=self.cwd, startupinfo=si)
	except Exception as ex:
		sublime.error_message(ErrorMessage.task_execution_failed(self.task_name, str(ex)))
```
, but it needs to be tested on Windows.

mgutski commented:
Closing due to prolonged inactivity and unresolved breaking changes. All relevant bug fixes and features from this PR have been included in v1.1.0.
